lib/improver/tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py

```diff
@@ -0,0 +1,643 @@
+                [1., 1., 0.4, 1., 1.],
+                [1., 1., 1., 1., 1.]]],
+              [[[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]]],
+              [[[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]]]]])
+        percentiles = np.array([10, 50, 90])
+        radius = 2000.
+        result = (
+            GeneratePercentilesFromACircularNeighbourhood(
+                percentiles=percentiles).run(
+                    self.cube, radius))
+        self.assertArrayAlmostEqual(result.data, expected)
+
+    def test_multi_point_multitimes(self):
+        """Test behaviour for points over multiple times."""
+        cube = set_up_cube(
+            zero_point_indices=((0, 0, 2, 2), (0, 1, 2, 1)), num_time_points=2,
+            num_grid_points=5)
+        expected = np.array(
+            [[[[[1., 1., 1., 1., 1.],
+                [1., 1., 0.4, 1., 1.],
+                [1., 0.4, 0.4, 0.4, 1.],
+                [1., 1., 0.4, 1., 1.],
+                [1., 1., 1., 1., 1.]],
+               [[1., 1., 1., 1., 1.],
+                [1., 0.4, 1., 1., 1.],
+                [0.4, 0.4, 0.4, 1., 1.],
+                [1., 0.4, 1., 1., 1.],
+                [1., 1., 1., 1., 1.]]],
+              [[[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]],
+               [[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]]],
+              [[[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]],
+               [[1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.],
+                [1., 1., 1., 1., 1.]]]]])
+        percentiles = np.array([10, 50, 90])
+        radius = 2000.
+        result = (
+            GeneratePercentilesFromACircularNeighbourhood(
+                percentiles=percentiles).run(
+                    cube, radius))
+        self.assertArrayAlmostEqual(result.data, expected)
+
+    def test_single_point_lat_long(self):
+        """Test behaviour for a single grid cell on lat long grid."""
+        cube = set_up_cube_lat_long()
+        msg = "Invalid grid: projection_x/y coords required"
+        radius = 6000.
+        with self.assertRaisesRegexp(ValueError, msg):
+            GeneratePercentilesFromACircularNeighbourhood().run(cube, radius)
+
+    def test_single_point_masked_to_null(self):
```

Comment:
My comment from the last review still stands:
_"This is masking a point with valid data in, not a point with NaNs in. The code does something quite different if there is a NaN in the place you masked, it actually spreads the NaN over the the circular neighbourhood. I think we need to think about this, as this is not the required behaviour, although this could be work for a follow on ticket."_

Neither behaviour is great, but both are going to be hard to change. So I'm not sure what to do.

Reply:
Yes, I think the handling of masked data isn't great, although this is wider problem. We probably need to agree more widely how to handle this.